libp2p/protocols/pubsub/floodsub.nim

```diff
@@ -63,50 +63,38 @@ method unsubscribePeer*(f: FloodSub, peer: PeerID) =
 
 method rpcHandler*(f: FloodSub,
                    peer: PubSubPeer,
-                   rpcMsgs: seq[RPCMsg]) {.async.} =
-  await procCall PubSub(f).rpcHandler(peer, rpcMsgs)
-
-  for m in rpcMsgs:                                  # for all RPC messages
-    if m.messages.len > 0:                           # if there are any messages
-      var toSendPeers = initHashSet[PubSubPeer]()
-      for msg in m.messages:                         # for every message
-        let msgId = f.msgIdProvider(msg)
-        logScope: msgId
-
-        if msgId notin f.seen:
-          f.seen.put(msgId)                          # add the message to the seen cache
-
-          if f.verifySignature and not msg.verify(peer.peerId):
-            trace "dropping message due to failed signature verification"
-            continue
-
-          if not (await f.validate(msg)):
-            trace "dropping message due to failed validation"
-            continue
-
-          for t in msg.topicIDs:                     # for every topic in the message
-            if t in f.floodsub:
-              toSendPeers.incl(f.floodsub[t])        # get all the peers interested in this topic
-            if t in f.topics:                        # check that we're subscribed to it
-              for h in f.topics[t].handler:
-                trace "calling handler for message", topicId = t,
-                                                     localPeer = f.peerInfo.id,
-                                                     fromPeer = msg.fromPeer.pretty
-
-                try:
-                  await h(t, msg.data)                 # trigger user provided handler
-                except CancelledError as exc:
-                  raise exc
-                except CatchableError as exc:
-                  trace "exception in message handler", exc = exc.msg
-
-        # forward the message to all peers interested in it
-        let published = await f.broadcast(
-          toSeq(toSendPeers),
-          RPCMsg(messages: m.messages),
-          DefaultSendTimeout)
-
-        trace "forwared message to peers", peers = published
+                   rpcMsg: RPCMsg) {.async.} =
+  await procCall PubSub(f).rpcHandler(peer, rpcMsg)
+
+  if rpcMsg.messages.len > 0:                      # if there are any messages
```

Comment:
This assumes that there will only be one `RPCMsg` per transmission, but that's an assumption that is very easy to break by a client that sends multiple `RPCMsg` messages per transmission, which wouldn't be a direct protocol violation. 

I checked `decodeRpcMsg` and it will only decode one `RPCMsg` per call and ignores any other possible `RPCMsg` messages that could be read from the pipe. This seems wrong as well, and can lead to silent packet/message drops.

Reply:
shouldn't each of these be separately length-prefixed?

Reply:
Not necessarily, an implementation can choose to batch several messages and in that case we would only ever process the first `RPCMsg`.

Reply:
actually, that's not possible to encode: protobuf is a stream of key-value, there's no end-of-message marker so the length prefix is the only way to do it

Reply:
Each message in libp2p is itself length prefixed so it is possible to batch several protobuf packets in one single libp2p message.

Reply:
ok - that's the part I don't understand - [here](https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/pubsub/pubsubpeer.nim#L88) we read one length-prefixed chunk of data and decode it as an rpc msg - based on that, `data` cannot contain more than one message because there's no additional delimiter?

Reply:
also, it looks like RPC messages are independent of each other - even if they were batched somehow, they should be processed one-by-one - I think it's easier to reason about the processing done here when it's structured this way - then any looping is done outside this function which clarifies the exception handling (which is hard to reason about otherwise, mid-loop)

Reply:
> ok - that's the part I don't understand - here we read one length-prefixed chunk of data and decode it as an rpc msg - based on that, data cannot contain more than one message because there's no additional delimiter?

Yes, we read the length prefixed data from the channel, but then we only read the first protobuf message from it, if the buffer contains more rpc messages, those would get ignored - I think this is a bug and should be handled. 

When we read and decode the first `RPCMsg`, we should additionally return the length that was consumed from the buffer by the `decodeRpcMsg` and if there are more bytes left, we should parse it again and so on - essentially, we should call `decodeRpcMsg` in a loop until the entire buffer has been consumed.

Reply:
>  I think it's easier to reason about the processing done here when it's structured this way

Yeah, this was just the place which hinted at this issue, but I agree that it should be kept as is here and the looping should be done at the `decodeRpcMsg` call site, or even inside it.

Reply:
> return the length that was consumed

that's the thing - protobuf always consumes all bytes - the wire encoding is just a stream of unordered key-values for the fields, there's no specific end

Reply:
```                                                                                               
 +---------------------------------------------------------------------------------+                 
 |      +----------------------+ +----------------------+ +----------------------+ |                 
 |buf   |                      | |                      | |                      | |                 
 |len   |       RPCMsg 1       | |       RPCMsg 2       | |        RPCMsg 3      | |                 
 |      +----------------------+ +----------------------+ +----------------------+ |                 
 +---------------------------------------------------------------------------------+                 
 ```

Yes, but like in the diagram above, if the buffer contains several `RPCMsg`'s, `decodeRpcMsg` will only parse the first one, if the buffer contains more messages, we won't ever return them from it. This is what `decodeRpcMsg` does:

```nim
proc decodeRpcMsg*(msg: seq[byte]): ProtoResult[RPCMsg] {.inline.} =
  trace "decodeRpcMsg: decoding message", msg = msg.shortLog()
  var pb = initProtoBuffer(msg)
  var rpcMsg: RPCMsg
  rpcMsg.messages = ? pb.decodeMessages()
  rpcMsg.subscriptions = ? pb.decodeSubscriptions()
  rpcMsg.control = ? pb.decodeControl()
  ok(rpcMsg)
```

Reply:
ok, so how does it know that `RPCMsg 1` ends and `RPCMsg 2` begins? at which byte offset does that start? keep in mind that all of `decodeMessages`, `decodeSubscriptions` etc read _all_ data. see https://developers.google.com/protocol-buffers/docs/techniques#streaming

Reply:
OK, in that case we should be good, seems like it would not be possible to batch many `RPCMsg` messages without length prefixing, which means that `readLp` would read a message at a time.